### DISCUSSION TITLE
Is there a rust crate that can be used to evaluate bitcoin script?

### DISCUSSION THREAD

**Discussion by casey** [rust-bitcoin/rust-bitcoin#1313] [Q&A] [answered]
I'd like to write a simple command-line utility that evaluates scripts, for learning purposes. (Also, if such a command-line utility already exists, pointers to that would be great too.)

**Comment by TheBlueMatt** [comment, marked as answer]
You may want rust-bitcoinconsensus, which wraps the Bitcoin Core script engine.

**Reply by casey** [reply]
Nice, thank you!

**Comment by RCasatta** [comment]
> Also, if such a command-line utility already exists, pointers to that would be great too.

This is not rust-based but very good https://github.com/bitcoin-core/btcdeb

**Comment by sanket1729** [comment]
This is porbably not what you are looking for, but there is an [interpreter](https://docs.rs/miniscript/7.0.0/miniscript/interpreter/struct.Interpreter.html) in the miniscript library. 
- not a command line tool
- does not work for scripts that are not miniscript

It outputs an iterator over [SatisfiedConstraint](https://docs.rs/miniscript/7.0.0/miniscript/interpreter/enum.SatisfiedConstraint.html). Given a script, it outputs all keys, timelocks, and hashlocks that were used in the satisfaction for the given (mini)script.
